### PR TITLE
migration case studies lotka-volterra, rna-pulse

### PR DESCRIPTION
**Pull request by jjocham**

### CODE REVIEW

**Comment by jjocham** [comment]
@flo-schu 
Jetzt gibt es noch 3 AssertionErrors

der 1. sieht für mich Leien nach nem Rundungsfehler aus aber wir hatten ja gestern gesprochen dass die Werte sehr klein sein können.

2. eigentlich das selbe nur wundert mich das Desired nichtmal im selben shape ist.

3. auch relativ kleine abweichung

Ich kann mich in die jeweiligen testfunktionen reinarbeiten aber vielleicht hast du schon direkt ne intuition

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,13 @@ __pycache__
 .pytest_cache
 case_studies/*/results/
 case_studies/*/tests/
+case_studies/lotka_volterra_UDE_case_study/scripts/results
+case_studies/lotka_volterra_UDE_case_study/scripts/extrapolation
+case_studies/lotka_volterra_UDE_case_study/scripts/hyperparams
+case_studies/lotka_volterra_UDE_case_study/scripts/simple_cases
+case_studies/lotka_volterra_UDE_case_study/scripts/losses
+case_studies/lotka_volterra_UDE_case_study/scripts/failed_training
+case_studies/*/scripts/case_studies
```
Comment (flo-schu) [inline .gitignore]: It would be best to move the entire Lotka Volterra UDE case study, to a different submodule. Than also the .gitignore lines are no longer necessary, here.

We could also discuss, if it might not be more efficient to transfer the relevant UDE classes into the lotka volterra case study, this way, we don't blow up the package

File: .gitmodules [hunk 2]
```diff
@@ -1,4 +1,8 @@
 [submodule "case_studies/lotka_volterra_case_study"]
 	path = case_studies/lotka_volterra_case_study
 	url = git@github.com:flo-schu/lotka_volterra_case_study
+	branch = migration-pymob-6
+[submodule "case_studies/tktd_rna_pulse"]
+	path = case_studies/tktd_rna_pulse
+	url = https://github.com/flo-schu/tktd_rna_pulse.git
```
Comment (flo-schu) [inline .gitmodules]: I'm not sure, if I want to have TKTD RNA PULSE as a submodule of pymob. The only submodules needed for pymob are strictly those that are necessary for testing. If there are no tests that rely on tktd_rna_pulse, remove it

File: case_studies/lotka_volterra_case_study
Comment (flo-schu) [inline case_studies/lotka_volterra_case_study:1]: As mentioned before, the more I think about it, the more I would support the idea, of copying the relevant parts of the UDE case study that are required for testing to the Lotka-volterra case study. @jjocham Do you think this is feasible?

File: case_studies/tktd_rna_pulse
Comment (flo-schu) [inline case_studies/tktd_rna_pulse:1]: As mentioned, prefer to remove this as a submodule

File: pymob/inference/analysis.py
Comment (flo-schu) [inline pymob/inference/analysis.py:1]: Well done!

File: pymob/inference/numpyro_backend.py [hunk 3]
```diff
@@ -1392,7 +1393,40 @@ def nuts_posterior(self, mcmc, model, key, obs):
...
+
+        n_total = sum(counts.values())
+        weight = 1.0 / len(obs_vars)
+        scales: Dict[str, float] = {}
+        for name, n_obs in counts.items():
+            if n_total > 0 and n_obs:
+                scale = (n_total / n_obs) * weight
+            else:
+                scale = 1.0
+            scales[f"{prefix}{name}"] = float(scale)
```
Comment (flo-schu) [inline pymob/inference/numpyro_backend.py:1427]: @merkuns: Interesting function, seems useful. Where did you apply that?

File: pymob/sim/config/base.py [hunk 4]
```diff
@@ -223,7 +225,10 @@ def save(self, fp: Optional[str]=None, force: bool=False) -> None:
             by_alias=True, 
             mode="json", 
             exclude_none=True,
-            exclude={"case_study": {"output_path", "data_path", "root", "init_root", "default_settings_path"}}
+            exclude={
+                "case_study": {"output_path", "data_path", "root", "init_root", "default_settings_path"},
+                "inference_optax": {"loss_function"}
```
Comment (flo-schu) [inline pymob/sim/config/base.py:230]: why is loss_function excluded here? is that safe behavior? I.e. can the config be reproduced without it. this means loss_function should be a property or computed variable

File: pymob/sim/evaluator.py [hunk 5]
```diff
@@ -238,10 +239,20 @@ def __init__(
...
+                    import equinox as eqx
+                    if solver == UDESolver:
+                        model_params, model_solver = eqx.partition(self.model, eqx.is_array)             
 
                 self._solver = solver(
-                    model=self.model,
+                    model=model_solver,
                     post_processing=self.post_processing,
                     
                     coordinates=frozen_coordinates,
```
Comment (flo-schu) [inline pymob/sim/evaluator.py:258]: I would be extremely happy if we could harmonize UDE and normal ODE model solve here. Bsaically, using the same diffrax solver for each model, would make all other optimization/inference backends available also for UDE models, which would be a major facilitator for the usability of UDE models. I will find some time in the near future to consider, how this could work.

File: pymob/sim/evaluator.py [hunk 6]
```diff
@@ -347,9 +358,21 @@ def __call__(self, seed=None):
...
+            if isinstance(self._solver, UDESolver):
+                params, static = eqx.partition(self.model, eqx.is_array)
+                Y_ = self._solver(params, **self.parameters)
+            elif isinstance(self._solver, SolverBase):
+                Y_ = self._solver(**self.parameters)
+            else:
+                Y_  = self._solver(parameters=self.parameters, **self._signature)
+        elif isinstance(self._solver, SolverBase):
             Y_ = self._solver(**self.parameters)
-
```
Comment (flo-schu) [inline pymob/sim/evaluator.py:352]: This would be resolved with harmonization between UDE and ODE solvers

**Review by flo-schu — COMMENTED** [review]
Hi @jjocham; i've looked at your fixes and they seem well done. @merkuns @jjocham  I've added some larger refactoring, enhacement comments before we can go ahead and merge this.

The biggest issue that needs to be resolved is still the compatibility of UDE models with the other inference backends, most importantly numpyro

**Comment by flo-schu** [comment]
> @flo-schu Jetzt gibt es noch 3 AssertionErrors
> 
> der 1. sieht für mich Leien nach nem Rundungsfehler aus aber wir hatten ja gestern gesprochen dass die Werte sehr klein sein können.
> 
>     2. eigentlich das selbe nur wundert mich das Desired nichtmal im selben shape ist.
> 
>     3. auch relativ kleine abweichung
> 
> 
> Ich kann mich in die jeweiligen testfunktionen reinarbeiten aber vielleicht hast du schon direkt ne intuition

Ich hatte leider noch keine Gelegenheit dazu. Ich schaue mir die Probleme gegen mitte nächster Woche an. Davor habe ich noch Deadline Stress.

File: pymob/inference/numpyro_backend.py (hunk 3 above)
Comment (merkuns) [reply pymob/inference/numpyro_backend.py:1427]: I don't remember writing that, I think this is @jjocham's work.

File: pymob/sim/config/base.py (hunk 4 above)
Comment (merkuns) [reply pymob/sim/config/base.py:230]: I think I excluded that because NetCDF files can't contain functions but I'm not sure anymore. Maybe there is some way we can transform the loss functions into a format that can be stored and then transform it back when importing the config. I probably simply didn't deal with it because there were more pressing issues. 

As of now, when loading the config, the loss function would not be affected and the default quadratic loss function would be used for training.

File: pymob/sim/evaluator.py (hunk 5 above)
Comment (merkuns) [reply pymob/sim/evaluator.py:258]: I outlined all the necessary traits of a UDE solver in the JAX/Equinox universe in my master's thesis. From my point of view, it would be best to apply the model class logic applied to the UDE models to all Pymob models. This way, the UDESolver could be used by all models. I don't know how this would affect the other infering algorithms though.

File: pymob/inference/numpyro_backend.py (hunk 3 above)
Comment (jjocham) [reply pymob/inference/numpyro_backend.py:1427]: @flo-schu 
Thats a helper function I needed for resolving an UnexpectedTracer Error in the tktd_pulse notebook after migrating pymob to np 2.x

Here's the respective [Commit](https://github.com/flo-schu/tktd_rna_pulse/commit/6f074807fe4e7b9cebe5f5e712d60aa8283deffd)

File: case_studies/lotka_volterra_case_study
Comment (jjocham) [reply case_studies/lotka_volterra_case_study:1]: @flo-schu
Sure, I could just copy the relevant files/file parts of the UDE case study into the normal case study package.

If I'm not mistaken, there are no pytest files so far that actually test the results.
As of now, the test5.ipynb file is the file that @merkuns told me is my sanity check that everything should work.

File: .gitmodules (hunk 2 above)
Comment (jjocham) [reply .gitmodules]: All right, I'll remove it from submodules within the next few days

File: pymob/sim/evaluator.py (hunk 5 above)
Comment (jjocham) [reply pymob/sim/evaluator.py:258]: what do you @flo-schu think about that?

How big would the changes be?

File: pymob/inference/numpyro_backend.py (hunk 3 above)
Comment (flo-schu) [reply pymob/inference/numpyro_backend.py:1427]: Thanks for clarifying. Looks good. Thanks for generalizing that 👍️

File: pymob/sim/config/base.py (hunk 4 above)
Comment (flo-schu) [reply pymob/sim/config/base.py:230]: Thanks for clarifying. This needs to be resolved---possibly by using model classes, which have the loss function built-in.

File: pymob/sim/evaluator.py (hunk 5 above)
Comment (flo-schu) [reply pymob/sim/evaluator.py:258]: I think using model classes would be the way to go, but imo this is a separate and very large issue, because it touches many of the core design decisions made in pymob. For merging the PR, I think I can live with the duplicate implementation. 

Further explanation:

we already use model classes already in mempyguts https://gitlab.uni-osnabrueck.de/memuos/mempyguts and I've already defined a protocol for what a model class should include in a child package derived from mempyguts https://gitlab.uni-osnabrueck.de/fschunck/guts_base/-/blob/main/guts_base/mod.py?ref_type=heads#L17-52

The way model classes are currently treated is that they are translated into the respective config sections (params_info -> model_parameters, state_variables -> data_structure) and rhs and likelihood_function (error function) are used directly. Inside pymob, the components (rhs, state variables, parameters, error model) are treated separately, and I would like to keep it that way.

With some other colleagues, I'm currently thinking along these lines. And we're developing protocols for defining models and data for parameter estimation. These definitions could be used for exactly this purpose.

This issue is related to a number of other issues / technical debt accumulated in the development of pymob. Therefore I would not like to do something premature here.

- https://github.com/flo-schu/pymob/issues/92
- https://github.com/flo-schu/pymob/issues/71
- https://github.com/flo-schu/pymob/issues/22
- https://github.com/flo-schu/pymob/issues/48

File: pymob/solvers/diffrax.py [hunk 7]
```diff
@@ -22,6 +23,13 @@
     RecursiveCheckpointAdjoint,
     LinearInterpolation,
 )
+equinox = import_optional_dependency(
+    "equinox", errors="raise", extra="set_inferer(backend='equinox') was not executed successfully, because "
+    "'equinox' dependencies were not found. They can be installed with "
+    "pip install pymob[equinox]. Alternatively:"
+)
+if equinox is not None:
+    import equinox as eqx
```
Comment (flo-schu) [inline pymob/solvers/diffrax.py:33]: Since equinox is a dependency of diffrax, I think it's safe to import this here. Can be removed.

File: pymob/solvers/diffrax.py [hunk 8]
```diff
@@ -261,3 +269,357 @@ def odesolve_splitargs(self, *args, odestates, n_odeargs, n_ppargs, n_xin):
...
+        '''
+        n_odestates = len(odestates)
+        y0 = args[:n_odestates]
+        ppargs = args[n_odestates:n_odestates+n_ppargs]
+        x_in = args[n_odestates+n_ppargs:n_odestates+n_ppargs+n_xin]
+        sol, interp = self.odesolve(model_params=model_params, y0=y0, x_in=x_in)
+        
+        res_dict = OrderedDict({v:val for v, val in zip(odestates, sol)})
+
+        return self.post_processing(res_dict, jnp.array(self.x), interp, *ppargs)
```
Comment (flo-schu) [inline pymob/solvers/diffrax.py:515]: This needs to be merged with the JaxSolver. I'm happy to use the modern decorators `@eqx.filter_jit` and abandon the homebrew jaxsolver, which is written as a dataclass to ensure non-mutability. @merkuns, could you list the differences between UDESolver and JaxSolver? I would then try to merge the two and check if tests continue passing.

File: pymob/solvers/diffrax.py [hunk 9]
```diff
@@ -261,3 +269,357 @@ def odesolve_splitargs(self, *args, odestates, n_odeargs, n_ppargs, n_xin):
...
+            saveat=diffrax.SaveAt(ts=ts),
+            event=diffrax.Event(cond_fn),
+            max_steps=int(self.max_steps),
+            throw = self.throw_exception
+        )
+
+        # Change shape of results
+        sol_y = tuple([sol.ys[:,i] for i in jnp.arange(sol.ys.shape[1])])
+        
+        return sol_y
```
Comment (flo-schu) [inline pymob/solvers/diffrax.py:625]: @merkuns is this still required, when the normal evaluator logic works for `UDEBase`?

File: pymob/sim/evaluator.py [hunk 10]
```diff
@@ -238,10 +239,20 @@ def __init__(
                 })
 
                 solver_options.update(solver_extra_options)
-                
+
+                model_solver = self.model
+
+                equinox = import_optional_dependency(
+                    "equinox", errors="ignore"
+                )
```
Comment (flo-schu) [inline pymob/sim/evaluator.py:247]: This should test for the diffrax (because equinox is a dependency of diffrax).

**Comment by flo-schu** [comment]
I've created a new PR #200 pointing on this branch for the architecture changes and harmonization between UDESolver, Jaxsolver and Evaluator

File: pymob/solvers/diffrax.py (hunk 8 above)
Comment (merkuns) [reply pymob/solvers/diffrax.py:515]: - The model is split into two parts: the parts that can and must be JITted (i.e., the model parameters) and the parts that can't and mustn't be JITted
- Conversely, because the UDE model function does not receive the model parameters as arguments, all references to `ode_args` were removed. That would have to be reversed when we are merging the two model types. I think that I actually wrote some code at some point that accounted for `ode_args` being empty. But I then decided to do it the other way (don't ask why, I don't remember anyway). Maybe I can find this code somewhere within my GIT commits.
- Because of the split between model parameters and the unJITtable parts, I could replace `partial(jax.jit, static_argnames=["self"])` with the much more convenient `eqx.filter_jit` (detailed explanation: master's thesis p. 26f.)
- The `x_in` part of the `args` passed to the diffrax solver has to either be a `LinearInterpolation` or an empty tuple because the `UDEBase` currently requires that. The `UDESolver` ensures that this is always the case. (lines 421-448)
- Equinox models have a certain way with input and output structures. I dealt with that in and after the call to `diffrax.diffeqsolve` (lines 475, 488) but it can definitely be done in a smarter way
- Temporal resolution of the results was matched to the data (line 474). This is important for training.

File: pymob/solvers/diffrax.py (hunk 9 above)
Comment (merkuns) [reply pymob/solvers/diffrax.py:625]: I think there were two reasons for this workaround:

1. Repeated calls to the evaluator didn't work on my (Windows) computer. There seems to be some kind of tracer leak that only appeared on my laptop (with the corresponding test failing, too) and did not show up on your (Linux) computer nor the Linux/Windows/MacOS tests on GitHub. Circumventing the evaluator and calling the `standalone_solver` directly solved that problem.
2. Also, this enabled me to efficiently implement batched training sets by simply passing the data batches to the standalone solver via an input argument, thereby staying within the JIT context. I was under the impression that within the Pymob logic, I would have had to meddle with the observations stored within an xArray which is unfortunately incompatible with JIT and would have slowed down training immensely. 

Getting rid of this workaround is probably the most complex challenge of harmonization. I can try to help you in some way but I always felt that this is beyond my abilities.
